### PR TITLE
fix: add C3 library support

### PR DESCRIPTION
**Pull request by id3nom**
fix(c3_lsp): update root directory to be able to handle project in the following order - a C3 project 'project.json', a library '.c3l' dir, and ancestor 'git' repo.

### CODE REVIEW

File: lua/lspconfig/server_configurations/c3_lsp.lua [hunk 1]
```diff
@@ -4,7 +4,21 @@ return {
   default_config = {
     cmd = { 'c3lsp' },
     root_dir = function(fname)
-      return util.root_pattern { 'project.json', '.git' }(fname)
+      --look for a project directory
+      local path = util.root_pattern {"project.json"}(fname);
```
Comment (justinmk) [inline lua/lspconfig/server_configurations/c3_lsp.lua:8]: why not do this? 
```suggestion
      local path = util.root_pattern {'project.json', 'manifest.json'}(fname);
```

File: lua/lspconfig/server_configurations/c3_lsp.lua [hunk 2]
```diff
@@ -4,7 +4,14 @@ return {
...
     root_dir = function(fname)
-      return util.root_pattern { 'project.json', '.git' }(fname)
+      --look for a project directory
+      local path = util.root_pattern { 'project.json', 'manifest.json' }(fname)
+      if path ~= nil then
+        return path
+      end
+
+      --look for .git directory
+      return util.find_git_ancestors(fname)
```
Comment (glepnir) [inline lua/lspconfig/server_configurations/c3_lsp.lua:14]:
```suggestion
      return util.root_patttern { 'project.json', 'manifest.json', '.git' }(fname)
```

Comment (id3nom) [reply lua/lspconfig/server_configurations/c3_lsp.lua:14]: @glepnir will this also work if we are editing the branch inside the '.git' directory instead?

Comment (glepnir) [reply lua/lspconfig/server_configurations/c3_lsp.lua:14]: this is an order will find every pattern until find root

File: lua/lspconfig/server_configurations/c3_lsp.lua [hunk 3]
```diff
@@ -4,7 +4,7 @@ return {
   default_config = {
     cmd = { 'c3lsp' },
     root_dir = function(fname)
-      return util.root_pattern { 'project.json', '.git' }(fname)
+      return util.root_patttern { 'project.json', 'manifest.json', '.git' }(fname)
```
Comment (justinmk) [inline lua/lspconfig/server_configurations/c3_lsp.lua:7]: this has a typo

Comment (glepnir) [reply lua/lspconfig/server_configurations/c3_lsp.lua:7]: oops sry.

**Comment by justinmk** [comment]
The [original form](https://github.com/neovim/nvim-lspconfig/pull/3309#discussion_r1766846520) of this PR is a hint that we may need to clean up more of the existing configs, since I assume that was copied from some other config.

**Comment by glepnir** [comment]
This might take a while to clean up, as there are a lot of files. I'll come back here after I've prioritized C's Pr.

**Comment by id3nom** [comment]
> The [original form](https://github.com/neovim/nvim-lspconfig/pull/3309#discussion_r1766846520) of this PR is a hint that we may need to clean up more of the existing configs, since I assume that was copied from some other config.

Yes indeed, it was based on existing configurations. tanks for your help @glepnir  and @justinmk

**Pull request merged**